x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APMTracer.java

```diff
@@ -0,0 +1,388 @@
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.xpack.apm.APMAgentSettings.APM_ENABLED_SETTING;
+import static org.elasticsearch.xpack.apm.APMAgentSettings.APM_TRACING_NAMES_EXCLUDE_SETTING;
+import static org.elasticsearch.xpack.apm.APMAgentSettings.APM_TRACING_NAMES_INCLUDE_SETTING;
+
+/**
+ * This is an implementation of the {@link org.elasticsearch.tracing.Tracer} interface, which uses
+ * the OpenTelemetry API to capture spans.
+ * <p>
+ * This module doesn't provide an implementation of the OTel API. Normally that would mean that the
+ * API's default, no-op implementation would be used. However, when the APM Java is attached, it
+ * intercepts the {@link GlobalOpenTelemetry} class and provides its own implementation instead.
+ */
+public class APMTracer extends AbstractLifecycleComponent implements org.elasticsearch.tracing.Tracer {
+
+    private static final Logger LOGGER = LogManager.getLogger(APMTracer.class);
+
+    /** Holds in-flight span information. */
+    private final Map<String, Context> spans = ConcurrentCollections.newConcurrentMap();
+
+    private volatile boolean enabled;
+    private volatile APMServices services;
+
+    private List<String> includeNames;
+    private List<String> excludeNames;
+    /** Built using {@link #includeNames} and {@link #excludeNames}, and filters out spans based on their name. */
+    private volatile CharacterRunAutomaton filterAutomaton;
+    private String clusterName;
+    private String nodeName;
+
+    public void setClusterName(String clusterName) {
+        this.clusterName = clusterName;
+    }
+
+    public void setNodeName(String nodeName) {
+        this.nodeName = nodeName;
+    }
+
+    /**
+     * This class is used to make all OpenTelemetry services visible at once
+     */
+    record APMServices(Tracer tracer, OpenTelemetry openTelemetry) {}
+
+    public APMTracer(Settings settings) {
+        this.includeNames = APM_TRACING_NAMES_INCLUDE_SETTING.get(settings);
+        this.excludeNames = APM_TRACING_NAMES_EXCLUDE_SETTING.get(settings);
+        this.filterAutomaton = buildAutomaton(includeNames, excludeNames);
+        this.enabled = APM_ENABLED_SETTING.get(settings);
+    }
+
+    void setEnabled(boolean enabled) {
+        this.enabled = enabled;
+        if (enabled) {
+            this.services = createApmServices();
+        } else {
+            destroyApmServices();
+        }
+    }
+
+    void setIncludeNames(List<String> includeNames) {
+        this.includeNames = includeNames;
+        this.filterAutomaton = buildAutomaton(includeNames, excludeNames);
+    }
+
+    void setExcludeNames(List<String> excludeNames) {
+        this.excludeNames = excludeNames;
+        this.filterAutomaton = buildAutomaton(includeNames, excludeNames);
+    }
+
+    @Override
+    protected void doStart() {
+        if (enabled) {
+            this.services = createApmServices();
```

Comment:
I think there might be a possible race condition here. If one thread was doing `setEnabled(false)` and we had already read `enabled` here we might leave APMServices as an instance with enabled false.

Reply:
I see what you're getting at, but in reality `doStart()` will only be called from `start()`, which is called from `Node.start()` in a single thread. There shouldn't be any cluster state modifications going on at this point that would cause `enabled` to change value.

Reply:
OK, cool.